Support segmented reductions and null mask reductions

isVoid opened the pull request:
closes #9135 
closes #9552 

This PR adds support for numeric types to `simple_op`, `sum`, `prod`, `min`, `max`, `any`, `all`. Also, this PR adds `segmented_null_mask_reduction` to compute null mask reductions on segments.

jrhemstad commented on cpp/src/reductions/simple.cuh:
```diff
@@ -46,29 +53,43 @@ namespace simple {
-std::unique_ptr<scalar> simple_reduction(column_view const& col,
-                                         rmm::cuda_stream_view stream,
-                                         rmm::mr::device_memory_resource* mr)
+scalar_or_column_ptr simple_reduction(
+  column_view const& col,
+  std::optional<std::reference_wrapper<column_view const>> offsets,
```
I'm having trouble supporting the idea of a segmented reduction and a normal reduction sharing the same code path. They are different algorithms with different implementations. This feels like a single responsibility violation at best and plain confusing at worst.

vyasr replied:
Yeah now that I'm seeing the practical application that you had in mind during our discussion I am also not convinced. Reducing LOC is nice, but these are fundamentally different and all you're gaining is to avoid duplicating a few levels of dispatch.

ttnghia commented:
I didn't read the implementation yet. But I assume that this is extremely similar to what sort-based groupby aggregate is doing. If this new implementation is more efficient than the current groupby aggregate implementation, we also need to switch to use segmented reduction for groupby aggregate too. In order to support this idea, please add a benchmark for comparison between their performance.

bdice commented on cpp/include/cudf/reduction.hpp:
```diff
@@ -69,6 +71,52 @@ std::unique_ptr<scalar> reduce(
+ * @param col Input column view
+ * @param agg Aggregation operator applied by the reduction
+ * @param offsets Indices to segment boundaries
+ * @param output_dtype  The computation and output precision.
```
Reductions are currently not being performed in the output precision. See issue #9988.

vyasr reviewed (approved):
CMake LGTM. Does using separate TUs for each reduction improve compile time a lot, or did you just do that to match previously existing reductions?

isVoid commented:
> CMake LGTM. Does using separate TUs for each reduction improve compile time a lot, or did you just do that to match previously existing reductions?

Just matching existing reductions.

isVoid commented:
@vyasr I added a benchmark for the segment reductions, some more changes to CMake for benchmarks, may require a second look. I added `cudftestutil` as one of the libraries to link to as it makes debugging benchmark code easier. Let me know if this makes sense.

isVoid commented:
@ttnghia When I looked at groupby cases, I think per group result is computed using `thrust::reduce_by_key`, where each "segments" is specified by neighboring group labels, while segmented reduction specifies inputs with segment boundaries. Though you can always convert one to another - I'm assuming the process adds non-trivial overheads. Thus I don't think one can be converted to another..

ttnghia commented:
> @ttnghia When I looked at groupby cases, I think per group result is computed using `thrust::reduce_by_key`, where each "segments" is specified by neighboring group labels, while segmented reduction specifies inputs with segment boundaries. Though you can always convert one to another - I'm assuming the process adds non-trivial overheads. Thus I don't think one can be converted to another..

At high level, segmented reduction seems to be more efficient to me but I'm not sure what is in reality. Different approaches have different load-balancing behaviors, and different overheads. Therefore, as mentioned above, a benchmark comparing two approaches is always the best way to defend the new way to the public.

bdice commented on cpp/src/reductions/simple_segmented.cuh:
```diff
@@ -0,0 +1,271 @@
+
+ public:
+  template <typename ElementType,
+            std::enable_if_t<is_supported<ElementType>() &&
+                             not cudf::is_fixed_point<ElementType>()>* = nullptr>
```
It looks like the `is_supported` function should also exclude fixed point values, rather than excluding them separately in the SFINAE templates.

bdice commented on cpp/src/reductions/simple_segmented.cuh:
```diff
@@ -0,0 +1,271 @@
+ *
+ * This is used by operations sum(), product(), and sum_of_squares().
+ * It only supports numeric types. If the output type is not the
+ * same as the input type, an extra cast operation may incur.
```
```suggestion
 * same as the input type, an extra cast operation may occur.
```

bdice commented on cpp/src/reductions/simple_segmented.cuh:
```diff
@@ -0,0 +1,271 @@
+                                         rmm::mr::device_memory_resource* mr)
+  {
+    auto result =
+      simple_segmented_reduction<ElementType, double, Op>(col, offsets, null_handling, stream, mr);
```
We need to put a TODO referencing issue #9988.

bdice commented on cpp/src/reductions/simple_segmented.cuh:
```diff
@@ -0,0 +1,271 @@
+   * @tparam ElementType The input column type or key type.
+   * @param col Input column (must be numeric)
+   * @param offsets Indices to segment boundaries
+   * @param output_type Requested type of the scalar result
```
Missing `@param null_handling`.

bdice reviewed (changes requested):
Round of suggestions attached. This is a big undertaking, so congrats on the progress so far! There were several comments that I only wrote once (or twice) but apply to the entire PR.

Please merge in changes from upstream as well - I have a feeling that clang-tidy compliance will require quite a few updates.

isVoid commented on cpp/include/cudf/detail/reduction.cuh:
```diff
@@ -218,6 +220,85 @@ std::unique_ptr<scalar> reduce(InputIterator d_in,
+          typename InputIterator,
+          typename OffsetIterator,
+          typename OutputType = typename thrust::iterator_value<InputIterator>::type,
+          typename std::enable_if_t<std::is_same_v<OutputType, string_view>>* = nullptr>
```
Currently segmented ops doesn't support fixed point yet. In `reduce` the `device_storage_type_t` were computed and the scale was computed as a post step. But I agree with what you suggest here.

isVoid commented on cpp/include/cudf/detail/reduction_functions.hpp:
```diff
@@ -254,5 +254,140 @@ std::unique_ptr<scalar> nth_element(
+/**
+ * @brief Compute sum of each segment in input column
+ *
+ * If all elements in input segment are null, the segment result is null.
```
I don't know, I kind of think this is an important detail to note. Considering the docstring for `segmented_reduce` is quite noisy in this context - if some user is looking to use this method specifically, perhaps it's most convenient to add this description here?

isVoid commented on cpp/include/cudf/detail/reduction_functions.hpp:
```diff
@@ -254,5 +254,140 @@ std::unique_ptr<scalar> nth_element(
+ *
+ * @param col Input column to compute any_of.
+ * @param offsets Indices to identify segment boundaries
+ * @param output_dtype Data type of return type and typecast elements of input column
```
Maybe? In `simple_segmented.cuh` this may help getting rid of `bool_resul_column_dispatcher`. I think the performance should be similar.

isVoid replied:
I noticed `any` is using `min` to implement, that means the result is being casted to bool regardless. To really support any `output_dtype` you would need an extra casting step. So ignore what I said above.

bdice replied:
I looked at this with @isVoid. We have to keep this `output_type` parameter because it is exposed to the user API. That way, we can show an error if the user requests an "all" or "any" reduction with a type besides bool8. While it's a little clunky, it is probably better to error than to cast the resulting `bool8` column to some other type like `double` or `int32`. Users needing to cast the result can do so explicitly.

isVoid commented on cpp/include/cudf/reduction.hpp:
```diff
@@ -69,6 +71,50 @@ std::unique_ptr<scalar> reduce(
+ * specified. If the input column has non-arithmetic type
+ *   eg.(timestamp, string...), the same type must be specified.
+ *
+ * If the reduction fails, the member is_valid of the output scalar
+ * will contain `false`.
```
Deferring to @karthikeyann, could you provide us with some insight about this line originally in the docstring of `reduce`? Thanks!

isVoid commented on cpp/include/cudf/reduction.hpp:
```diff
@@ -69,6 +71,50 @@ std::unique_ptr<scalar> reduce(
+ * will contain `false`.
+ *
+ * @param col Input column view
+ * @param agg Aggregation operator applied by the reduction
```
I believe offsets is included, but at the wrong line.

isVoid commented on cpp/include/cudf/reduction.hpp:
```diff
@@ -69,6 +71,50 @@ std::unique_ptr<scalar> reduce(
+ * output type is not bool8.
+ *
+ * If the input column has arithmetic type, output_dtype can be any arithmetic
+ * type. For `mean`, `var` and `std` ops, a floating point output type must be
+ * specified. If the input column has non-arithmetic type
```
That might be a right way to go. I don't like how part of the `output_dtype` constaints are mentioned in `@throw` and part are mentioned here. But let me first check it's actually going to throw first.

bdice replied:
Removed this for segmented reduce (these are not yet supported by `segmented_reduce`) and fixed the docstring for `reduce`.

isVoid commented on cpp/src/reductions/segmented_all.cu:
```diff
@@ -0,0 +1,47 @@
+  CUDF_EXPECTS(output_dtype == cudf::data_type(cudf::type_id::BOOL8),
+               "all() operation can be applied with output type `BOOL8` only");
+
+  // dispatch for non-dictionary types
```
Dictionary types are not supported for now. In `all.cu` there's a separate fast path for that.

isVoid commented on cpp/src/reductions/segmented_all.cu:
```diff
@@ -0,0 +1,47 @@
+#include "cudf/types.hpp"
+#include <cudf/detail/reduction_functions.hpp>
+#include <cudf/detail/utilities/device_atomics.cuh>
+#include <reductions/simple_segmented.cuh>
```
This is a local include. Note that `simple_segmented.cuh` is in the same folder as `segmented_*.cu`. Maybe we should simply write `#include "simple_segmented.cuh"`?

bdice replied:
I guess quotes would be preferred by this rule unless this header could be a “detail include”.

> Use quotes " to include local headers from the same relative source directory. This should only occur in source files and non-public header files. Otherwise use angle brackets <> around included header filenames.

— cuDF Developer Guide

isVoid commented on cpp/src/reductions/segmented_reductions.cpp:
```diff
@@ -0,0 +1,114 @@
+      case aggregation::ALL:
+        return reduction::segmented_all(col, offsets, output_dtype, null_handling, stream, mr);
+        break;
+      default: CUDF_FAIL("Unsupported aggregation type.");
```
Is it ok to add them in a separate PR? Those are `compound_ops` which takes some more scaffolding to the `simple_ops` introduced in this one.

bdice replied:
Definitely ok to do this later! Perhaps add a TODO indicating the future intent to add `compound_ops`?

isVoid commented on cpp/src/reductions/simple_segmented.cuh:
```diff
@@ -0,0 +1,271 @@
+  size_type num_segments = offsets.size() - 1;
+
+  auto result = [&] {
+    if (col.has_nulls()) {
```
I'd love to explore this as well.

isVoid replied:
We can follow up in a different PR.

isVoid commented on cpp/src/reductions/segmented_reductions.cpp:
```diff
@@ -0,0 +1,114 @@
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  // TODO: handle invalid inputs.
```
Can't remember either - this might be added outside of our pairing session. Perhaps referring to empty inputs handling like `reductions.cpp`?

bdice replied:
This can be resolved. We're changing this API to take `lists_column_view` and removing this comment.

vyasr commented:
> @vyasr I added a benchmark for the segment reductions, some more changes to CMake for benchmarks, may require a second look. I added `cudftestutil` as one of the libraries to link to as it makes debugging benchmark code easier. Let me know if this makes sense.

@isVoid for clarification, am I correct that you've added this target so that you'll be able to access `cudftestutil` functionality while debugging, but you don't expect those functions to be used any more once you've finalized the benchmarks? I don't see any fundamental issues with this and I think it's fine if you expect future benchmark writers to also benefit from having access to those functions during development.

isVoid commented:
> > @vyasr I added a benchmark for the segment reductions, some more changes to CMake for benchmarks, may require a second look. I added `cudftestutil` as one of the libraries to link to as it makes debugging benchmark code easier. Let me know if this makes sense.
> 
> @isVoid for clarification, am I correct that you've added this target so that you'll be able to access `cudftestutil` functionality while debugging, but you don't expect those functions to be used any more once you've finalized the benchmarks? I don't see any fundamental issues with this and I think it's fine if you expect future benchmark writers to also benefit from having access to those functions during development.

Yes, I kinda expected developers can benefit from helpers in testutils (especially `cudf::test::print`) during development/debugging. But once done it's not needed. So really it's a convenience thing.

vyasr reviewed (approved):
CMake approval

jrhemstad commented on cpp/include/cudf/reduction.hpp:
```diff
@@ -69,6 +73,44 @@ std::unique_ptr<scalar> reduce(
+ * @returns Output column with results of segmented reduction.
+ */
+std::unique_ptr<column> segmented_reduce(
+  lists_column_view const& col,
```
Remind me why this is a `lists_column_view` instead of separate offsets/data? This isn't really a "segmented" reduction anymore. It's a "per-list reduction".

bdice replied:
Good question! I briefly alluded to this during our libcudf meeting today but didn’t fully explain due to time. If you recall the conversations we had about indexing and iterators on #9588, there are two styles of indexing used in libcudf. One is what I call “N+1”, and requires the indices to be monotonic and contiguous. The endpoint of one segment is the start of the next segment. This is how lists and strings columns are indexed. The other style is what I call “2N”, where the start and endpoints are independent, and the segments can be arbitrarily overlapping. That is used for slicing, and is the format used by cub (which we call to perform the reduction).

The 2N indexing is most powerful when the index iterators are template parameters, allowing for arbitrary device iterators as we enabled for null mask reductions in #9588. We don’t want a template function in the public API, but that would be our preferred internal/detail implementation for maximum flexibility and alignment with the cub code that we call.

After discussion with @isVoid, we were unable to think of a second use case for the public API accepting a data column and a column of N+1 indices. The only use case we could identify was the list reduction, so a stronger commitment to the N+1 indexing scheme in a public API seemed pointless. Moreover, a public API accepting N+1 indices would need to be bounds-checked to ensure safety while the public API we settled on for a lists column reduction does not need to be checked since we know the list indexing is contiguous and monotonic by construction.

bdice replied:
Should we rename this API to `lists_reduce`?

jrhemstad replied:
> The 2N indexing is most powerful when the index iterators are template parameters

Agreed, luckily with transform iterators its pretty easy to translate `2N` <-> `N+1`. 

> we were unable to think of a second use case for the public API accepting a data column and a column of N+1 indices

Doing segmented reductions is quite normal in C++, so it only makes sense to me expose it in a canonical/generic way that can be used both for lists and just standard segmented reductions.

bdice replied:
Okay. Then these are the steps I'd suggest:
- API is `segmented_reduce(column_view data, column_view indices, ...)`, where `indices` are of the N+1 format
- Implement bounds-checking over indices to ensure they are monotonic/contiguous (we also need to document those requirements but we have to bounds-check anyway because we don't want to allow illegal memory accesses from public APIs)
- Consider whether to add a convenience method for `lists_reduce(lists_column_view, ...)`

Future PR:
- Refactor the detail/internal method to use 2N indexing with iterator template parameters

jrhemstad replied:
I would do:
```
segmented_reduce(column_view data, device_span<size_type> offsets, ...)
```

Likewise, you can have two overloads of `segmented_reduce` for `2N` and `N+1` similar to `split/slice`. 

> Implement bounds-checking over indices to ensure they are monotonic/contiguous (we also need to document those requirements but we have to bounds-check anyway because we don't want to allow illegal memory accesses from public APIs)

https://github.com/rapidsai/cudf/issues/5505

isVoid replied:
Changed to `device_span` at public API level for indices. Per last two comments above.

isVoid replied:
I've chatted with @bdice offline and we both thought making a `device_span` from a `column_view` is not particularly easy today. We were thinking of a convenience function for user. At one point we were thinking of `to_device_span` member method; but on a second thought, but perhaps a `make_device_span` factory function is better?

vyasr replied:
FYI Essentially the inverse operations is requested in #9656.

jrhemstad replied:
Just add an implicit conversion operation to `column_view` to `device_span<T>` that verifies the type of the `column_view` matches `T`.

bdice replied:
In progress in #10302.

bdice replied:
#10302 is now merged. To wrap up the previous pieces of the discussion:
- We will not implement bounds checking, but rather document the expectations of the API per #5505.
- We will not add a convenience method for `lists_reduce(lists_column_view, ...)` unless requested later. For now, @revans2 is fine with the APIs available in this PR and #10302.
- Follow-up work will refactor the detail/internal method to use 2N indexing with iterator template parameters.
I think that resolves this conversation.

jrhemstad commented on cpp/include/cudf/reduction.hpp:
```diff
@@ -69,6 +73,44 @@ std::unique_ptr<scalar> reduce(
+ */
+std::unique_ptr<column> segmented_reduce(
+  lists_column_view const& col,
+  std::unique_ptr<aggregation> const& agg,
```
```suggestion
  aggregation const& agg,
```

revans2 commented on cpp/include/cudf/reduction.hpp:
```diff
@@ -69,6 +72,46 @@ std::unique_ptr<scalar> reduce(
+ * the same output type must be specified.
+ *
+ * @param col Input column view
+ * @param offsets Indices to segment boundaries
```
Some more details here might be nice. Is this the same as the offsets in a list column? i.e. the first entry is always 0.

From the code it looks like the number of entries returned is determined by the number of offsets. Is there any issue if the offsets do not include everything in `col`? I assume not, but just want to be sure that the contract is clear.

What happens if the offsets are empty?  Will the resulting value be null? is it aggregation dependent?

isVoid replied:
> Is this the same as the offsets in a list column? i.e. the first entry is always 0.

This API expect input has at least 1 segment (1 row in the list column), the rest is the same with the usual offsets column.

> Is there any issue if the offsets do not include everything in col?

It's ok. Ultimately we use a pair of neighboring items in the offset column to specify begin and end range of elements to reduce for. As long as the indices doesn't go beyond the data boundary it should be well defined.

And this is a test case that's missing at the moment.

> What happens if the offsets are empty?

Do you mean if `offsets` column is empty? It's currently rejected as invalid input.

Will update docstrings accordingly.

isVoid replied:
> This API expect input has at least 1 segment (1 row in the list column), the rest is the same with the usual offsets column.

Revise - when you have 0 rows and 1 item in the offset vector (0), this is also valid input.

revans2 reviewed (approved):
The API looks good to me. I had a few questions that from reading through the code I was able to answer, but it would be nice to have them answered in the API docs itself so the contract it clear.

revans2 commented:
Oh I wanted to add that I will try to put together Java/Spark patches to use this and give feedback if I have any.

revans2 commented:
How exactly am I supposed to transform the offsets inside a `cudf::lists_column_view` into a `cudf::device_span<cudf::size_type>` so I can use it with this API? I can get access to the raw `data` pointer, and the `size`, but there is no `device_span` constructor that takes those as arguments.  I have to have a`thrust::device_vector`, an `rmm:device_vector` or an `rrm:device_uvector` to create this. The current API is not going to work for my use case unless I have a way to do this.

bdice commented:
> How exactly am I supposed to transform the offsets inside a `cudf::lists_column_view` into a `cudf::device_span<cudf::size_type>` so I can use it with this API? I can get access to the raw `data` pointer, and the `size`, but there is no `device_span` constructor that takes those as arguments. I have to have a`thrust::device_vector`, an `rmm:device_vector` or an `rrm:device_uvector` to create this. The current API is not going to work for my use case unless I have a way to do this.

I proposed a `lists_column_view` API above, see discussion thread here: https://github.com/rapidsai/cudf/pull/9621#discussion_r796237246

@jrhemstad's suggestion sounds like a path forward here:

> Just add an implicit conversion operation to `column_view` to `device_span<T>` that verifies the type of the `column_view` matches `T`. And that it's not nullable.

revans2 commented:
> > Just add an implicit conversion operation to `column_view` to `device_span<T>` that verifies the type of the `column_view` matches `T`. And that it's not nullable.

That sounds great. Should I do that or is that something that would be done here?

bdice commented:
> > > Just add an implicit conversion operation to `column_view` to `device_span<T>` that verifies the type of the `column_view` matches `T`. And that it's not nullable.
> 
> That sounds great. Should I do that or is that something that would be done here?

I'm working on this now, in a separate PR #10302.

isVoid commented:
I updated the docstring for `segmented_reduction`, to describe the specification of `offsets`. Also added a note on the current up-cast of precision when reducing integral and floating point values, and remove mentioning of `computation precision` at `output_dtype`.

revans2 reviewed (approved):
This combined with #10302 looks to give me what I need

bdice commented on cpp/src/reductions/segmented_reductions.cpp:
```diff
@@ -0,0 +1,103 @@
+                                         rmm::cuda_stream_view stream,
+                                         rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(offsets.size() > 0, "`offsets` vector should have at least 1 element.");
```
Since `offsets` are indexed as N+1 segments, a single element in `offsets` means there are zero segments. Does that mean the result is empty? If so, does the value of the single offset matter? Must it be `0` or `<= segmented_values.size()` or is there any bound on it at all?

isVoid replied:
> a single element in offsets means there are zero segments. Does that mean the result is empty?

That's correct. That single value must be 0, however we don't check it because of https://github.com/rapidsai/cudf/issues/5505. Out of bound behaviors are undefined. Like `gather` if `DONT_CHECK` flag is specified. Part of this was copied from `segmented_shift.cu`.

I think we can start putting together a specification for `offsets` derived from [arrow columnar format](https://arrow.apache.org/docs/format/Columnar.html#variable-size-binary-layout). Thinking about situations of below:
- `length==0` case
- Can `offsets` only cover part of the data column?
- What's the behavior of out of bound indices?

isVoid replied:
I mentions undefined behavior of out of bound indices in docstring.

bdice reviewed (approved):
I have a few minor comments but I think this is sufficiently ready for merge. We have a few TODO items that we should probably track in an issue.

isVoid commented on cpp/tests/reductions/segmented_reduction_tests.cpp:
```diff
@@ -0,0 +1,353 @@
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(*res, expect);
+}
+
+// TODO: Add tests for null_policy::INCLUDE
```
`null_policy::INCLUDE` tests was indeed added. I forgot to remove the TODO.

jrhemstad commented on cpp/include/cudf/detail/null_mask.cuh:
```diff
@@ -501,6 +505,110 @@ std::vector<size_type> segmented_null_count(bitmask_type const* bitmask,
 }
 
+/**
+ * @brief Reduce an input null mask using segments defined by offset indices
+ * into an output null mask.
```
Reduced how? What does it mean to "reduce" a null mask?

isVoid replied:
It's arguably "reduction" is a bit too broad for this operation. But it might be easier for user of the function to understand the net outcome of the function as a fold with binary `AND` of all the bits if `null_handling` is `INCLUDE`, `OR` otherwise. (Though implementation doesn't really do that).

bdice replied:
Beyond the `AND`/`OR`, there's special behavior for handling empty segments. This is hard to succinctly describe, and perhaps a generic `@brief` followed by a more extensive description is the best we can do. I've tried to come up with a better way to write this but haven't come up with anything I like.

bdice replied:
Maybe:
```suggestion
 * @brief Create an output null mask whose validity is determined by the
 * validity of any/all elements of segments of an input null mask.
```

jrhemstad commented on cpp/include/cudf/detail/null_mask.cuh:
```diff
@@ -501,6 +505,110 @@ std::vector<size_type> segmented_null_count(bitmask_type const* bitmask,
+  // Empty segments are always null in the output mask
+  auto const num_segments =
+    static_cast<size_type>(std::distance(first_bit_indices_begin, first_bit_indices_end));
+  auto [output_null_mask, output_null_count] = cudf::detail::valid_if(
+    segment_length_iterator,
+    segment_length_iterator + num_segments,
+    [] __device__(auto const& len) { return len > 0; },
+    stream,
+    mr);
```
It seems wasteful to first compute a null mask for empty segments and another for segments w/ or w/o nulls and AND them together. Let's fuse those steps.

```suggestion
  auto const segment_valid_counts = segmented_count_bints(...);
  auto const length_and_valid_count = thrust::zip_iterator{segment_length_iterator, segment_valid_counts.begin()};
  auto [output_null_mask, output_null_count] = cudf::detail::valid_if(
    length_and_valid_count,
    length_and_valid_count + num_segments,
    [] __device__(auto const& length_and_valid_count) { 
       auto const length = thrust::get<0>(length_and_valid_count);
       auto const valid_count = thrust::get<1>(length_and_valid_count);
       return (length > 0) and (null_handling == null_policy::EXCLUDE) ?  valid_count > 0 : valid_count == segment_length; 
    },
    stream,
    mr);
```

jrhemstad commented on cpp/include/cudf/reduction.hpp:
```diff
@@ -69,6 +72,56 @@ std::unique_ptr<scalar> reduce(
+std::unique_ptr<column> segmented_reduce(
+  column_view const& segmented_values,
+  device_span<size_type const> offsets,
+  aggregation const& agg,
```
yes, I believe @nvdbaranec is working on updating `reduce` to take a `reduce_aggregation`.

nvdbaranec replied:
Correct.  PR for this should be up in a day or so.

bdice commented:
@gpucibot merge

bdice commented:
@gpucibot merge
